src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManagerImpl.java

```diff
@@ -0,0 +1,227 @@
+        Lifecycle lifecycle = Lifecycle.NONE;
+
+        TransactionImpl(int nestingDepth) {
+            this.nestingDepth = nestingDepth;
+        }
+
+        void validateCurrentTransaction() {
+            validateCurrentThread();
+
+            if (currentDepth == -1 || stack.get(currentDepth) != this) {
+                String errorMessage = String.format(
+                        "Transaction function was called on a transaction with depth %d, but the current transaction has depth %d.",
+                        nestingDepth,
+                        currentDepth);
+                throw new IllegalStateException(errorMessage);
+            }
+        }
+
+        // Validate that this transaction is open.
+        private void validateOpen() {
+            if (lifecycle != Lifecycle.OPEN) {
+                throw new IllegalStateException("Transaction operation cannot be applied to a closed transaction.");
+            }
+        }
+
+        @Override
+        public Transaction openNested() {
+            validateCurrentTransaction();
+            validateOpen();
+            return open();
+        }
+
+        private void close(Result result) {
+            validateCurrentTransaction();
+            validateOpen();
+            // Block transaction operations
+            lifecycle = Lifecycle.CLOSING;
+
+            // Note: it is important that we don't let exceptions corrupt the global state of the transaction manager.
+            // That is why any callback has to run inside a try block.
+            RuntimeException closeException = null;
+
+            // Invoke callbacks in reverse order
+            for (int i = closeCallbacks.size()-1; i >= 0; i--) {
+                try {
+                    closeCallbacks.get(i).onClose(this, result);
+                } catch (Exception exception) {
+                    if (closeException == null) {
+                        closeException = new RuntimeException("Encountered an exception while invoking a transaction close callback.", exception);
+                    } else {
+                        closeException.addSuppressed(exception);
+                    }
+                }
+            }
+
+            closeCallbacks.clear();
+
+            if (currentDepth == 0) {
+                lifecycle = Lifecycle.OUTER_CLOSING;
+
+                // Invoke outer close callbacks in reverse order
+                for (int i = outerCloseCallbacks.size() - 1; i >= 0; i--) {
+                    try {
+                        outerCloseCallbacks.get(i).afterOuterClose(result);
+                    } catch (Exception exception) {
+                        if (closeException == null) {
+                            closeException = new RuntimeException("Encountered an exception while invoking a transaction outer close callback.", exception);
+                        } else {
+                            closeException.addSuppressed(exception);
```

Comment:
In the case where an exception is encountered in both a close callback and the outer callback, do we want to somehow specify that instead of just having it get added directly as being suppressed with the exception message pointing at the close callback instead of the outer close callback?